compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -958,14 +959,26 @@ LogicalResult DataTiledMMAAttr::populateOperandOffsetsSizesStrides(
       }));
 
   // Populate tile offsets by delinearizing threadId over the CrossThread dims.
-  SmallVector<OpFoldResult> tileOffsetsBasis = getAsIndexOpFoldResult(
-      ctx, sliceSwizzledShape(swizzle, [](TileSwizzle::Dim d) {
+  // Since the AffineDelinearizeIndexOp does not bound the input index, we
+  // must bound the threadId by the product of the offset ranges.
+  SmallVector<int64_t> tileOffsetsBasis =
+      sliceSwizzledShape(swizzle, [](TileSwizzle::Dim d) {
         return d.kind == TileSwizzle::Dim::Kind::CrossThread;
-      }));
+      });
+
+  // Bound for threadId is the product of tileOffsetsBasis.
+  OpFoldResult threadIdBound =
+      builder.getIndexAttr(ShapedType::getNumElements(tileOffsetsBasis));
+  AffineExpr d0 = builder.getAffineDimExpr(0), d1 = builder.getAffineDimExpr(1);
+  OpFoldResult boundedThreadId = affine::makeComposedFoldedAffineApply(
+      builder, loc, {d0 % d1}, {threadId, threadIdBound});
```

Comment:
this can be just a `builder.create<arith::RemSIOp>`.

Reply:
I chose to go with an affine apply because I think it makes canonicalizations easier (canonicalizing affine maps with other affine maps). Maybe I'm wrong, though, and the canonicalize patterns already take into account arith ops?

Reply:
I have no idea! Feel free to leave it like this. I'm going to start now writing missing folding patterns for whatever is missing, so it doesn't make a big difference for us either way. I was thus more focusing on whatever makes our C++ code shorter.

Reply:
yep, I think affine ops are better for those index calculation. We'd have some chances to cancel some calculation, hopefully.